integration/run_test.go

```diff
@@ -0,0 +1,128 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/util/testutil"
+)
+
+// TestRunMultipleContainersWithSameImage runs two containers with the same image at the same time
+func TestRunMultipleContainersWithSameImage(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, nginxImage})
+}
+
+// TestRunMultipleContainersWithDifferentImages runs two containers with different images at the same time
+func TestRunMultipleContainersWithDifferentImages(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, drupalImage})
+}
+
+func RunMultipleWebContainers(t *testing.T, imageNames []string) {
+	regConfig := newRegistryConfig()
+	sh, done := newShellWithRegistry(t, regConfig)
+	defer done()
+
+	// Initialize config files for containerd and snapshotter
+	additionalConfig := ""
+	if !isTestingBuiltinSnapshotter() {
+		additionalConfig = proxySnapshotterConfig
+	}
+	containerdConfigYaml := testutil.ApplyTextTemplate(t, `
+version = 2
+
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+{{.AdditionalConfig}}
+`, struct {
+		AdditionalConfig string
+	}{
+		AdditionalConfig: additionalConfig,
+	})
+
+	// Setup environment
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, []byte(containerdConfigYaml), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	sh.
+		X("update-ca-certificates").
+		Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+	rebootContainerd(t, sh, "", "")
+	for _, imageName := range deduplicate(imageNames) {
+		// Mirror image
+		copyImage(sh, dockerhub(imageName), regConfig.mirror(imageName))
+		indexDigest := optimizeImage(sh, regConfig.mirror(imageName))
+
+		// Pull image
+		sh.X("ctr", "i", "pull", "--user", regConfig.creds(), regConfig.mirror(imageName).ref)
+
+		// Create SOCI metadata
+		sh.X("soci", "create", regConfig.mirror(imageName).ref)
+		sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(imageName).ref)
```

Comment:
You already pulled the image in line 95. So this `rpull` will just skip the lazy-loading since the image already exists in the local storage.

Reply:
Reaching out about this elsewhere.

Reply:
Not exactly. `containerd` and `soci-snapshotter` do not share the content store. What will happen, is that for each layer in container image, there will be a respective call to `Prepare()`, which will go over mounting the layer depending on Soci index configuration.

Reply:
@vkuzniet I think that's the behavior we want here. Can you clarify if you're also suggesting some change?

Reply:
They do not share the content store for artifacts, but share it for images.

Reply:
> @vkuzniet I think that's the behavior we want here. Can you clarify if you're also suggesting some change?

I see. If the intention of the test is to lazily pull the container image twice, then I'm fine with that.

Reply:
> They do not share the content store for artifacts, but share it for images.

For download/decompress part, snapshotter explicitly uses the same content store as for artifacts. See https://github.com/awslabs/soci-snapshotter/blob/main/fs/unpacker.go#L76 for more details. Snapshotter does not have an access to containerd's content addressable store and we didn't have an intention to make this happen in the first place. Anyway, if the call to rpull lands on Prepare(), snapshotter will take over and will attempt to lazily load the layer to the best of its ability (more details here: https://github.com/awslabs/soci-snapshotter/blob/main/snapshot/snapshot.go#L263).

Reply:
Per standup and post-scrum conversations, here is a manual recreation of the critical steps of this test, outside of a containerized environment. I believe this shows that the `ctr i pull` then `soci image rpull` produces the desired behavior with two of the layers being served by soci through the fuse mounted filesystems.

```
% docker run -d -p 5000:5000 --restart=always --name registry ghcr.io/oci-playground/registry:v3.0.0-alpha.1

12ce889755d9b883e941223c84e041f92a2ab5e32ab5ccbdbd0230ff2615cc0d

% curl http://localhost:5000/v2/_catalog

{"repositories":[]}

% sudo ctr i pull docker.io/library/nginx:latest                      
docker.io/library/nginx:latest:                                                   resolved       |++++++++++++++++++++++++++++++++++++++| 
index-sha256:0047b729188a15da49380d9506d65959cce6d40291ccfb4e039f5dc7efd33286:    done           |++++++++++++++++++++++++++++++++++++++| 
manifest-sha256:9a821cadb1b13cb782ec66445325045b2213459008a41c72d8d87cde94b33c8c: done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:5f63362a3fa390a685ae42e1936feeca3e4fba185bdc46fb66cf184036611f7d:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:50c68654b16f458108a537c9842c609f647a022fbc5a9b6bde1ffb60b77c2349:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:88d3ab68332da2aa6cc8d83c9dfe95905dc899d9b8fb302ebae2bf9a6b167c40:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:40b838968eeab5abc9fb941a8e3ee1377660bb02672153cada52bc9d4e0595b7:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:3f4ca61aafcd4fc07267a105067db35c0f0ac630e1970f3cd0c7bf552780e985:    done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:1403e55ab369cd1c8039c34e6b4d47ca40bbde39c371254c7cba14756f472f52:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:3ed295c083ec7246873f1b98bbc7b634899c99f6d2d901e2f9f5220d871830dd:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.7 s                                                                    total:  52.3 M (74.6 MiB/s)                                      
unpacking linux/amd64 sha256:0047b729188a15da49380d9506d65959cce6d40291ccfb4e039f5dc7efd33286...
done: 1.266077552s

% sudo ctr i tag docker.io/library/nginx:latest localhost:5000/nginx:latest
localhost:5000/nginx:latest

% sudo ctr i push --platform linux/amd64 --plain-http localhost:5000/nginx:latest   
manifest-sha256:9a821cadb1b13cb782ec66445325045b2213459008a41c72d8d87cde94b33c8c: done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:1403e55ab369cd1c8039c34e6b4d47ca40bbde39c371254c7cba14756f472f52:   done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.4 s                                                                    total:  9.0 Ki (22.5 KiB/s)                                      

% curl http://localhost:5000/v2/_catalog

{"repositories":["nginx"]}

% sudo soci create localhost:5000/nginx:latest
layer sha256:5f63362a3fa390a685ae42e1936feeca3e4fba185bdc46fb66cf184036611f7d -> ztoc skipped
layer sha256:3ed295c083ec7246873f1b98bbc7b634899c99f6d2d901e2f9f5220d871830dd -> ztoc skipped
layer sha256:40b838968eeab5abc9fb941a8e3ee1377660bb02672153cada52bc9d4e0595b7 -> ztoc skipped
layer sha256:88d3ab68332da2aa6cc8d83c9dfe95905dc899d9b8fb302ebae2bf9a6b167c40 -> ztoc skipped
layer sha256:50c68654b16f458108a537c9842c609f647a022fbc5a9b6bde1ffb60b77c2349 -> ztoc sha256:678ea3b10a0dfab41fd7dfdc39596485f0e213e067043d43aa9f1d74458516ff
layer sha256:3f4ca61aafcd4fc07267a105067db35c0f0ac630e1970f3cd0c7bf552780e985 -> ztoc sha256:1c5fe514fa61890c0f1684fdbdf8a14f61d3a61bbed1c503cee5a7887f25e0ca

% sudo soci index list
DIGEST                                                                     SIZE    IMAGE REF                         PLATFORM
sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd    1083    docker.io/library/nginx:latest    linux/amd64    
sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd    1083    localhost:5000/nginx:latest       linux/amd64    

% sudo soci push --plain-http localhost:5000/nginx:latest
pushing soci index with digest: sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd
pushing artifact with digest: sha256:678ea3b10a0dfab41fd7dfdc39596485f0e213e067043d43aa9f1d74458516ff
pushing artifact with digest: sha256:1c5fe514fa61890c0f1684fdbdf8a14f61d3a61bbed1c503cee5a7887f25e0ca
skipped artifact with digest: sha256:9a821cadb1b13cb782ec66445325045b2213459008a41c72d8d87cde94b33c8c
successfully pushed artifact with digest: sha256:1c5fe514fa61890c0f1684fdbdf8a14f61d3a61bbed1c503cee5a7887f25e0ca
successfully pushed artifact with digest: sha256:678ea3b10a0dfab41fd7dfdc39596485f0e213e067043d43aa9f1d74458516ff
pushing artifact with digest: sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd
successfully pushed artifact with digest: sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd

% sudo soci-snapshotter-grpc&> ~/soci-snapshotter-logs &
[1] 27651

% sudo soci image rpull --plain-http --soci-index-digest sha256:cff81469d225a23cd8aefa7641279029def23325a02b6406a6a1f54c9b642ecd localho
st:5000/nginx:latest
fetching sha256:9a821cad... application/vnd.docker.distribution.manifest.v2+json
fetching sha256:1403e55a... application/vnd.docker.container.image.v1+json
fetching sha256:3ed295c0... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:88d3ab68... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:40b83896... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:5f63362a... application/vnd.docker.image.rootfs.diff.tar.gzip

% mount   
[boilerplate output omitted]
fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
overlay on /local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/merged type overlay (rw,relatime,lowerdir=/local/docker/overlay2/l/DL3KS6KPL25RMZ3PVT56XFK7ER:/local/docker/overlay2/l/C4OAMNIB7AF56EUYPNS2BOMDBR:/local/docker/overlay2/l/SXOW7OH6WBV6RG5LXPF5TCRBCS:/local/docker/overlay2/l/DEMTSB4QMCQCRCRSSSNDAAAAI2:/local/docker/overlay2/l/TZSZV6EQADHJUIAYDQ7UDCBM4V,upperdir=/local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/diff,workdir=/local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/work)
nsfs on /run/docker/netns/b773e06fa9e8 type nsfs (rw)
/local/home/clrnc/src/soci-snapshotter/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
/local/home/clrnc/src/soci-snapshotter/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/2/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)

% sudo ctr run -d --rm --snapshotter soci --net-host localhost:5000/nginx:latest nginxExample

% sudo ctr c ls
CONTAINER       IMAGE                          RUNTIME                  
nginxExample    localhost:5000/nginx:latest    io.containerd.runc.v2    

% sudo ctr t ls
TASK            PID      STATUS    
nginxExample    29651    RUNNING

% mount
[boilerplate output omitted]
fusectl on /sys/fs/fuse/connections type fusectl (rw,relatime)
overlay on /local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/merged type overlay (rw,relatime,lowerdir=/local/docker/overlay2/l/DL3KS6KPL25RMZ3PVT56XFK7ER:/local/docker/overlay2/l/C4OAMNIB7AF56EUYPNS2BOMDBR:/local/docker/overlay2/l/SXOW7OH6WBV6RG5LXPF5TCRBCS:/local/docker/overlay2/l/DEMTSB4QMCQCRCRSSSNDAAAAI2:/local/docker/overlay2/l/TZSZV6EQADHJUIAYDQ7UDCBM4V,upperdir=/local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/diff,workdir=/local/docker/overlay2/e903856086088109af564197ad1752045447f628fa49b31aef83b09736b8ee6d/work)
nsfs on /run/docker/netns/b773e06fa9e8 type nsfs (rw)
/local/home/clrnc/src/soci-snapshotter/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
/local/home/clrnc/src/soci-snapshotter/soci on /var/lib/soci-snapshotter-grpc/snapshotter/snapshots/2/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
overlay on /run/containerd/io.containerd.runtime.v2.task/default/nginxExample/rootfs type overlay (rw,relatime,lowerdir=/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/6/fs:/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/5/fs:/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/4/fs:/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/3/fs:/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/2/fs:/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/1/fs,upperdir=/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/7/fs,workdir=/var/lib/soci-snapshotter-grpc/snapshotter/snapshots/7/work)
```